plugin/pkg/shoot/validator/admission.go

```diff
@@ -949,8 +953,12 @@ func validateMachineImagesConstraints(constraints []garden.MachineImage, image,
 	validValues := []string{}
 
 	for _, v := range constraints {
+		if v.ExpirationDate != nil && v.ExpirationDate.Time.UTC().Before(time.Now().UTC()) {
```

Comment:
I don't understand this code. It does not consider the `image` nor `oldImage`. What does it do?

Reply:
it loops over all machine images in the cloud profile to check if the image specified in the shoot is valid.

In order for a machine image to be valid for the shoot creation & update, the expiration date must not be in the past. Then if the name and version of the image from the CloudProfile matches the shoots new image, the image specified in the shoot is valid.

So the line above makes sure that a shoot cannot specify an image from the cloud profile with an expiration date in the past.